### ISSUE TITLE
support for custom key used for the HTTP11 Connetion Pool

### ISSUE THREAD

**Issue by GlenDC** [scrapy/scrapy#5129]
## Summary

Make it possible to use a custom key for each connection made within the http11 code,
or at least allow part of it to be configurable, currently these keys are non-configurable.

## Motivation

Relevant parts of code:

- https://github.com/scrapy/scrapy/blob/e27eff47ac9ae9a9b9c43426ebddd424615df50a/scrapy/core/downloader/handlers/http11.py#L217
- https://github.com/scrapy/scrapy/blob/e27eff47ac9ae9a9b9c43426ebddd424615df50a/scrapy/core/downloader/handlers/http11.py#L245

These are the only 2 places I found where we set the key in order to get a (TCP) connection.
This is great mostly, as it does already:

- cache connections in a pools;
- allows the reuse of the connections;
- does so in a logical way by taking into account the proxy configuration, schema and endpoint;

It does however not allow for any control on which connections to reuse.

As such it is currently not possible AFAIK —regardless if you use a proxy or not — to execute a series of requests on the same connection in a reliable manner within a concurrent environment. That is, if you have a chain of requests through 3 different parse functions let's say, then there is no way to be sure that all 3 run over the same connection.

This makes true HTTP/1.1 Keep-Alive Persistent connections possible.

It would be great if somehow:

- I could provide a callback that can be used to create such a key myself based on the request;
- or that there is built-in a way to actually ensure that a same connection for a follow-up request is used, similar to how Scrapy does allow the same cookiejar to be used between follow-up requests.

## Describe alternatives you've considered

- You could not use the core http11 functionality at all and use your own downloader code I Imagine. But not sure how would that work? Would that mean that I can switch to experimental Asyncio and do use it there? Or do I need to roll out my own reactor entirely? Or can I customise just parts of an existing reactor?
   - This I've considered but seemed a bit too big in scope and/or too experimental for something I wish to run reliable in a production environment;
- You could also limit the pool to a single connection I imagine (not sure how, bit I suppose it is possible) but that isn't desirable as it means that all my concurrent requests would need to run on a single request, despite actually being different http1.1 sessions (or at least they should be separate);

## Additional Context

Above are the alternatives I've considered, but none seem desirable.

That being said, I'm not an expert in Scrapy. I tried to search the internet, books, and the Scrapy codebase + documentation for answers. And as far as I can see and understand there is no way for me to actually ensure follow-up requests run on the same connection.

This however seems odd to me as that would mean that there is no real way in a concurrent environment to use True HTTP/1.1 Keep-Alive (so called persistent) connections? At least not if you want to guarantee that different grouped requests stay separated despite using the same host, schema and proxy router.

**Comment by wRAR** [comment]
Do you want the persistent connection to improve download speed, or?

**Comment by GlenDC** [comment]
No, I want it in order to be able to use the same connection with my internal proxy service as to be able to maintain my entire session, just as a regular browser would be able to maintain that same connection for all its requests for that one user session.

**Comment by GlenDC** [comment]
Found a way to do it by MonkeyPatching it, but it seems a bit ugly if you ask me:

```python
def scrapy_proxy_agent_request(self, method, uri, headers=None, bodyProducer=None):
    """
    Issue a new request via the configured proxy.
    """
    # group all connections by used proxy & session ID
    session_id = headers._rawHeaders.get(b'my-custom-session-id-header')
    if session_id:
        session_id = session_id[0].decode('utf-8')
    else:
        session_id = 'global'
    return self._requestWithEndpoint(
        key=("http-proxy", self._proxyURI.host, self._proxyURI.port, session_id),
        endpoint=self._getEndpoint(self._proxyURI),
        method=method,
        parsedURI=URI.fromBytes(uri),
        headers=headers,
        bodyProducer=bodyProducer,
        requestPath=uri,
    )


ScrapyProxyAgent.request = scrapy_proxy_agent_request
```

**Comment by Gallaecio** [comment]
If you want a cleaner approach until this is implemented in Scrapy itself, you could subclass the default download handler class for HTTP, and set it in [DOWNLOAD_HANDLERS](https://docs.scrapy.org/en/latest/topics/settings.html#download-handlers) to use it. You can then subclass `ScrapyProxyAgent` with your custom `request` method, and have your custom download handler class use your `ScrapyProxyAgent` subclass.

**Comment by GlenDC** [comment]
Thank you @Gallaecio for your input. I agree that your suggested approach is cleaner. Going to give it a try somewhere this week and I'll come back to you!

**Comment by AdrianAtZyte** [comment]
I am not convinced this is something we should implement in Scrapy, specially given the complexity to extend this to all supported download handlers and the "escape hatch" of being able to get this through a custom download handler.
